modules/percolator/src/main/java/org/elasticsearch/percolator/ExtractQueryTermsService.java

```diff
@@ -132,19 +150,29 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
                     longestTerm = term;
                 }
             }
-            return Collections.singleton(longestTerm);
+            return new Result(false, longestTerm);
         } else if (query instanceof BooleanQuery) {
-            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
-            boolean hasRequiredClauses = false;
+            BooleanQuery bq = (BooleanQuery) query;
+            List<BooleanClause> clauses = bq.clauses();
+            int mnsm = bq.getMinimumNumberShouldMatch();
+            int numRequiredClauses = 0;
+            int numOptionalClauses = 0;
+            int numProhibitedClauses = 0;
             for (BooleanClause clause : clauses) {
                 if (clause.isRequired()) {
-                    hasRequiredClauses = true;
-                    break;
+                    numRequiredClauses++;
+                }
+                if (clause.isProhibited()) {
+                   numProhibitedClauses++;
+                }
+                if (clause.isScoring()) {
+                    numOptionalClauses++;
                 }
             }
-            if (hasRequiredClauses) {
-                UnsupportedQueryException uqe = null;
+            if (numRequiredClauses > 0) {
+                boolean alwaysMatch = true;
```

Comment:
I think we should compute alwaysMatch in a different for-loop to keep things simple since it needs different information. (For instance I think the current impl would return alwaysMatch=true if there is a MUST clause with a match_all query and a MUST_NOT clause, which is wrong?)